### PR TITLE
agnoster: cut down on fork/execs improve performance when not in a repo

### PR DESCRIPTION
**Pull request by 0xabu**
This PR improves performance, especially when we're not in a git repo, by cutting down the number of forks/execs needed just to render the prompt. It's a particularly noticeable speedup on Cygwin.

There are three sub-commits here for three distinct performance bugs; any of them could be applied independently if you don't like some of the others.

I've also submitted similar changes upstream as https://github.com/agnoster/agnoster-zsh-theme/pull/44

### CODE REVIEW

**Comment by mcornella** [comment]
The `echo`s are not fork/exec's but `write()` (or `print()` which is basically the same). So using `PROMPT+=...` is not needed. It also creates a weird bug that makes my prompt be displayed one character to the right, leaving a space before it, until I restart the session with `exec zsh`. I haven't been able to find the culprit:

image

The downside of reverting that commit is that it makes it so that your `$jobstates` performance improvement doesn't return correct results. I suppose in PROMPT land we cannot access that variable.

File: themes/agnoster.zsh-theme [hunk 1]
```diff
@@ -55,22 +56,22 @@ prompt_segment() {
...
   if [[ -n $CURRENT_BG ]]; then
-    echo -n " %{%k%F{$CURRENT_BG}%}$SEGMENT_SEPARATOR"
+    PROMPT+=" %{%k%F{$CURRENT_BG}%}$SEGMENT_SEPARATOR"
   else
-    echo -n "%{%k%}"
+    PROMPT+="%{%k%}"
   fi
-  echo -n "%{%f%}"
+  PROMPT+="%{%f%}"
   CURRENT_BG=''
```
Comment (mcornella) [inline themes/agnoster.zsh-theme]: I found the bug: this should be `CURRENT_BG='NONE'` so that next prompt begins with the right background.

**Comment by 0xabu** [comment]
@mcornella re. "The echos are not fork/exec's but write()" -- I realise that, but you still do need a subshell (which is one fork/exec) to capture the output of the echo.

I will make the change you suggested -- thanks for catching this!

**Comment by 0xabu** [comment]
Gentle ping @mcornella -- I believe the last update addressed all of your PR comments, and I also tested with Mercurial.

**Comment by mcornella** [comment]
Thanks!

**Pull request merged**